### DISCUSSION TITLE
Best method to the get the package name of a loaded Rego file?

### DISCUSSION THREAD

**Discussion by lquerel** [microsoft/regorus#228] [Q&A] [answered]
Is there any particular reason for the `get_modules` method in `Engine` to be annotated with `#[doc(hidden)]`? I have a use case where I need to determine the package name declared in a Rego file. Initially, I thought to obtain this information from the `add_policy_from_file` method, but it does not return any reference to the module. I then noticed the public `get_modules` method, but it has this annotation.

**Comment by anakrish** [comment, marked as answer]
The reason [get_modules](https://github.com/microsoft/regorus/blob/69d64266638ff8ec65edcc8ec1a8b5453383fb85/src/engine.rs#L207) is hidden is that it exposes the AST which is unstable at this point in time. There are some changes being considered for the AST that will overall benefit Regorus, but might break someone that uses the AST.

The API exists in engine (https://github.com/microsoft/regorus/pull/50) so as to not break @eric-therond who uses the AST directly. The hidden API will likely be removed after a few releases, once that use case is also properly supported, without exposing the internals.

@thedavemarshall also [expressed](#187 ) interest in obtaining the list of loaded modules.

Maybe a good way is to have `add_policy` functions return the name of the package?

Currently, one way to obtain the name of the packages would be to evaluate a query that requests the keys within the data document
E.g:
```bash
$ regorus eval -d policy.rego "data[package_name]"
...
      "bindings": {
        "package_name": "agent_policy"
      }
...
```


```rust
let mut engine = Engine::new();
engine.add_policy(...);
engine.set_strict_builtin_errors(false); // Ensure that the query below does not raise errors due to missing data/input.

// Do not add_data since the added data shows up as keys in the `data` document.

// Create bindings to package names.
let results = engine.eval_query("data[package_name]".to_string(), false)?;
// Name of first package.
let package_name = results.result[0].bindings["package_name"].as_string()?.to_string();
```

**Reply by lquerel** [reply]
Thanks for this answer. Having the `add_policy` method returning the name of the package will be enough for my use case.
